### ISSUE TITLE
One-line parallel call...

### ISSUE THREAD

**Issue by mscuthbert** [joblib/joblib#544]
I absolutely love joblib's parallelization module.  So much easier than using multiprocessing directly -- should be in the standard library!

I've used it a lot, but I've found that maybe there's a need for an even easier interface for amateurs to be able to get parallel running working even easier than the easy way joblib already makes possible?  I've also had the need to run the same function directly and within an already parallelized function -- so in the latter case there's no speed gain and just more overhead.  I was wondering if the joblib team had any use for these pair of functions (with a little helper function) that might bring parallelizing
even closer to a simple drop-in.  Call like:

```python
def fib(n):
    # fib computing here
    return computedFib
fibNumbers = runParallel(fib, range(100))
```

It has some built in functionality for unpacking lists of tuples/lists, giving updates (or calling custom update code), and having a non-parallelizing version with identical function calls.

With updates, it is far less good than other solutions if the amount of time taken to do some inputs is very long compared to other parts, but for relatively even inputs (or relatively even among elements that are close in the iterable), it is nice to be able to get updates frequently.

Code is at https://github.com/cuthbertLab/music21/blob/master/music21/common/parallel.py
It's BSD-3-Clause licensed, so use if you'd like it.  Generated documentation at
http://web.mit.edu/music21/doc/moduleReference/moduleCommonParallel.html

The code (with one adaptation to make it Py2.7/3 compatible -- I'm now Py3 only) is below:

```python
import multiprocessing
from joblib import Parallel, delayed 

def runParallel(iterable, parallelFunction,
                updateFunction=None, updateMultiply=3,
                unpackIterable=False, updateSendsIterable=False):
    '''
    runs parallelFunction over iterable in parallel, optionally calling updateFunction after
    each cpus()  * updateMultiply calls.

    Setting updateMultiply too small can make it so that cores wait around when they
    could be working if one CPU has a particularly hard task.  Setting it too high
    can make it seem like the job has hung.

    updateFunction should take three arguments: the current position, the total to run,
    and the most recent results.  It does not need to be pickleable, and in fact,
    a bound method might be very useful here.  Or updateFunction can be "True"
    which just prints a generic message.

    If unpackIterable is True then each element in iterable is considered a list or
    tuple of different arguments to parallelFunction.

    If updateSendsIterable is True then the update function will get the iterable
    content, after the output.

    As of Python 3, partial functions are pickleable, so if you need to pass the same
    arguments to parallelFunction each time, make it a partial function before passing
    it to runParallel.

    Note that parallelFunction, iterable's contents, and the results of calling parallelFunction
    must all be pickleable, and that if pickling the contents or
    unpickling the results takes a lot of time, you won't get nearly the speedup
    from this function as you might expect.  
    
    >>> files = ['bach/bwv66.6', 'schoenberg/opus19', 'AcaciaReel']
    >>> def countNotes(fn):
    ...     c = corpus.parse(fn)
    ...     return len(c.recurse().notes)
    >>> #_DOCS_SHOW outputs = runParallel(files, countNotes)
    >>> outputs = common.runNonParallel(files, countNotes) #_DOCS_HIDE can't pickle doctest funcs.
    >>> outputs
    [165, 50, 131]
    
    Set updateFunction=True to get an update every 3 * numCpus (-1 if > 2)
    
    >>> #_DOCS_SHOW outputs = common.runParallel(files, countNotes, updateFunction=True)
    >>> outputs = common.runNonParallel(files, countNotes, updateFunction=True) #_DOCS_HIDE
    Done 0 tasks of 3
    Done 3 tasks of 3
    
    With a custom updateFunction that gets each output:
    
    >>> def yak(position, length, output):
    ...     print("%d:%d %d is a lot of notes!" % (position, length, output))
    >>> #_DOCS_SHOW outputs = common.runParallel(files, countNotes, updateFunction=yak)
    >>> outputs = common.runNonParallel(files, countNotes, updateFunction=yak) #_DOCS_HIDE
    0:3 165 is a lot of notes!
    1:3 50 is a lot of notes!
    2:3 131 is a lot of notes!    
    
    Or with updateSendsIterable, we can get the original files data as well:
    
    >>> def yik(position, length, output, fn):
    ...     print("%d:%d (%s) %d is a lot of notes!" % (position, length, fn, output))
    >>> #_DOCS_SHOW outputs = common.runParallel(files, countNotes, updateFunction=yik,
    >>> outputs = common.runNonParallel(files, countNotes, updateFunction=yik, #_DOCS_HIDE
    ...             updateSendsIterable=True)
    0:3 (bach/bwv66.6) 165 is a lot of notes!
    1:3 (schoenberg/opus19) 50 is a lot of notes!
    2:3 (AcaciaReel) 131 is a lot of notes!

    unpackIterable is useful for when you need to send multiple values to your function
    call as separate arguments.  For instance, something like:
    
    >>> def pitchesAbove(fn, minPitch): # a two-argument function
    ...     c = corpus.parse(fn) # again, the slow call goes in the function
    ...     return len([p for p in c.pitches if p.ps > minPitch])
    
    >>> inputs = [('bach/bwv66.6', 60),
    ...           ('schoenberg/opus19', 72),
    ...           ('AcaciaReel', 66)]
    >>> #_DOCS_SHOW outputs = common.runParallel(inputs, pitchesAbove, unpackIterable=True)
    >>> outputs = common.runNonParallel(inputs, pitchesAbove, unpackIterable=True) #_DOCS_HIDE
    >>> outputs
    [99, 11, 123]
    '''
    # multiprocessing has trouble with introspection
    # pylint: disable=not-callable
    if multiprocessing.current_process().daemon:
        return runNonParallel(iterable, parallelFunction, 
                              updateFunction=updateFunction,
                              updateMultiply=updateMultiply, 
                              unpackIterable=unpackIterable,
                              updateSendsIterable=updateSendsIterable)

    iterLength = len(iterable)
    totalRun = 0
    numCpus = cpus()
    if updateFunction is None:
        updateMultiply = iterLength 
        # if there is no need for updates, run at max speed 
        #    -- do the whole list at once.
    resultsList = []
    
    def callUpdate(ii):
        if updateFunction is True:
            print("Done {} tasks of {}".format(min([ii, iterLength]), 
                                               iterLength))
        elif updateFunction not in (False, None):
            for thisPosition in range(ii - (updateMultiply * numCpus), ii):
                if thisPosition < 0:
                    continue
                
                if thisPosition >= len(resultsList):
                    thisResult = None
                else:
                    thisResult = resultsList[thisPosition]
                
                if updateSendsIterable is False:
                    updateFunction(thisPosition, iterLength, thisResult)
                else:
                    updateFunction(thisPosition, iterLength, thisResult, iterable[thisPosition])

    callUpdate(0)    

    with Parallel(n_jobs=numCpus) as para:
        delayFunction = delayed(parallelFunction)
        while totalRun < iterLength:
            endPosition = min(totalRun + numCpus * updateMultiply, iterLength)
            rangeGen = range(totalRun, endPosition)

            if unpackIterable:
                _r = para(delayFunction(*iterable[i]) for i in rangeGen)
            else:
                _r = para(delayFunction(iterable[i]) for i in rangeGen)

            totalRun = endPosition
            resultsList.extend(_r)
            callUpdate(totalRun)

    return resultsList


def runNonParallel(iterable, parallelFunction,
                updateFunction=None, updateMultiply=3,
                unpackIterable=False, updateSendsIterable=False):
    '''
    This is intended to be a perfect drop in replacement for runParallel, except that
    it runs on one core only, and not in parallel.

    Used automatically if we're already in a parallelized function.
    '''
    iterLength = len(iterable)
    resultsList = []

    def callUpdate(ii):
        if ii % updateMultiply != 0:
            return
        
        if updateFunction is True:
            print("Done {} tasks of {}".format(min([ii, iterLength]), 
                                               iterLength))
        elif updateFunction not in (False, None):
            for thisPosition in range(ii - updateMultiply, ii):
                if thisPosition < 0:
                    continue
                
                if thisPosition >= len(resultsList):
                    thisResult = None
                else:
                    thisResult = resultsList[thisPosition]
                
                if updateSendsIterable is False:
                    updateFunction(thisPosition, iterLength, thisResult)
                else:
                    updateFunction(thisPosition, iterLength, thisResult, iterable[thisPosition])


    callUpdate(0)
    for i in range(iterLength):
        if unpackIterable:
            _r = parallelFunction(*iterable[i])
        else:
            _r = parallelFunction(iterable[i])

        resultsList.append(_r)

        callUpdate(i + 1)

    return resultsList


def cpus():
    '''
    Returns the number of CPUs or if >= 3, one less (to leave something out for multiprocessing)
    '''
    cpuCount = multiprocessing.cpu_count()
    if cpuCount >= 3:
        return cpuCount - 1
    else:
        return cpuCount

```

**Comment by lesteve** [comment]
Thanks a lot for the kind words and taking the time to open an issue! It feels great to hear that you find joblib useful and that you built something of value on top of it to solve some of the problems you are facing.

I did not have time to look at all the materials you sent in detail but here is my 2 cents about it:
* it is rather unlikely that we are going to provide a thin wrapper on top of `Parallel` like yours. I am guessing that for some users a function is easier to use that a class. We are well aware of this but this is still the interface we went for.
* some of the functionality you have may be a work-around for something that can be improved in joblib. It that is the case, then I would be more than happy if you could open separate issue with a more focused scope to see if there is something we can do about it. As an example is that that the `verbose`  argument of `Parallel` is not good enough for your purposes and that you want more details for reporting progress. Although not super high priority, this is an improvement that could well be considered. I seem to remember there was an issue in scikit-learn with GridSearchCV that was producing not so great output so that could be an argument for this as well.

Other comments:
* it feels like you are making assumptions that may be violated and that also that may lead to suboptimal code, for example, splitting the iterable in chunks. Quick jobs can be grouped in batches in order to suffer less from the multiprocessing overhead.
* following PEP8 coding conventions may help you attract users/developers.

**Comment by lesteve** [comment]
And by the way, the field of application (i.e. musicology if I understand correctly) seems like a really fun one!

**Comment by mscuthbert** [comment]
Thanks for the comments!  And support.  I just wouldn't feel like I was being a good member of the community if I didn't offer back what I found most useful.  Feel free to close, but maybe someone will find this useful

Yes, the project was originally a Perl project and got a bit of a community there, and our main methods mirror Javascript DOM methods (getElementsByClass etc.) so naming of methods and attributes follows the camelCase standard that came from those communities.

**Comment by lesteve** [comment]
> Thanks for the comments! And support. I just wouldn't feel like I was being a good member of the community if I didn't offer back what I found most useful. Feel free to close, but maybe someone will find this useful

Thanks a bunch for the community effort and taking the time to open this issue. I am going to close this issue, people will still able to find the issue through Google. As I said above feel free to open smaller scope issues if you have the spare bandwidth. Although probably not very high priority, improving how joblib reports progress could be a good candidate for a smaller scope issue. Even better than issues, opening short focussed PRs is the best way to get changes in joblib ;-).

> Yes, the project was originally a Perl project and got a bit of a community there, and our main methods mirror Javascript DOM methods (getElementsByClass etc.) so naming of methods and attributes follows the camelCase standard that came from those communities. Since most of our musician users come to programming for the first time in order to use music21 (not in order to use python specifically), lack of PEP8 naming hasn't been a detriment to attracting users. But we follow PEP8 in most other respects.

Makes complete sense. Practicality beats purity as they say :-).
